### DISCUSSION TITLE
ANR

### DISCUSSION THREAD

**Discussion by 474246820** [Tencent/libpag#2216] [Q&A] [answered]
使用版本com.tencent.tav:libpag:4.3.47

  #00  pc 0x0000000000054070  /apex/com.android.runtime/lib64/bionic/libc.so (abort+164)
  #01  pc 0x0000000000055d18  /apex/com.android.runtime/lib64/bionic/libc.so (__fortify_fatal(char const*, ...)+124)
  #02  pc 0x00000000000b9cb4  /apex/com.android.runtime/lib64/bionic/libc.so (HandleUsingDestroyedMutex(pthread_mutex_t*, char const*)+60)
  #03  pc 0x00000000000b9b44  /apex/com.android.runtime/lib64/bionic/libc.so (pthread_mutex_lock+240)
  #04  pc 0x0000000000230abc  /data/app/~~wsz9HcFa10baIQ8osrAiyA==/com.aimo.vika-Z8MMV_tqyIbL2M4UK-1bSw==/split_config.arm64_v8a.apk!libpag.so (BuildId: c43d6dd0367e51b4744d4c7eafb275d4782fb666)
  #05  pc 0x00000000000e0cec  /data/app/~~wsz9HcFa10baIQ8osrAiyA==/com.aimo.vika-Z8MMV_tqyIbL2M4UK-1bSw==/split_config.arm64_v8a.apk!libpag.so (pag::PAGLayer::duration() const+68) (BuildId: c43d6dd0367e51b4744d4c7eafb275d4782fb666)
  #06  pc 0x000000000010a778  /data/app/~~wsz9HcFa10baIQ8osrAiyA==/com.aimo.vika-Z8MMV_tqyIbL2M4UK-1bSw==/split_config.arm64_v8a.apk!libpag.so (Java_org_libpag_PAGLayer_duration+68) (BuildId: c43d6dd0367e51b4744d4c7eafb275d4782fb666)
  #07  pc 0x0000000000409654  /data/app/~~wsz9HcFa10baIQ8osrAiyA==/com.aimo.vika-Z8MMV_tqyIbL2M4UK-1bSw==/oat/arm64/base.odex (art_jni_trampoline+116)
  #08  pc 0x0000000000566d5c  /data/app/~~wsz9HcFa10baIQ8osrAiyA==/com.aimo.vika-Z8MMV_tqyIbL2M4UK-1bSw==/oat/arm64/base.odex (org.libpag.PAGImageView.onAnimationUpdate+92)
  #09  pc 0x000000000055d0a0  /data/app/~~wsz9HcFa10baIQ8osrAiyA==/com.aimo.vika-Z8MMV_tqyIbL2M4UK-1bSw==/oat/arm64/base.odex (org.libpag.PAGAnimator.onAnimationUpdate+144)
  #10  pc 0x000000000033eba4  /apex/com.android.art/lib64/libart.so (art_quick_invoke_stub+612)
  #11  pc 0x000000000033cc28  /apex/com.android.art/lib64/libart.so (art::JValue art::InvokeVirtualOrInterfaceWithVarArgs<art::ArtMethod*>(art::ScopedObjectAccessAlreadyRunnable const&, _jobject*, art::ArtMethod*, std::__va_list)+772)
  #12  pc 0x0000000000562cbc  /apex/com.android.art/lib64/libart.so (art::JNI<false>::CallVoidMethodV(_JNIEnv*, _jobject*, _jmethodID*, std::__va_list)+192)
  #13  pc 0x00000000001003b4  /data/app/~~wsz9HcFa10baIQ8osrAiyA==/com.aimo.vika-Z8MMV_tqyIbL2M4UK-1bSw==/split_config.arm64_v8a.apk!libpag.so (BuildId: c43d6dd0367e51b4744d4c7eafb275d4782fb666)
  #14  pc 0x0000000000103e28  /data/app/~~wsz9HcFa10baIQ8osrAiyA==/com.aimo.vika-Z8MMV_tqyIbL2M4UK-1bSw==/split_config.arm64_v8a.apk!libpag.so (BuildId: c43d6dd0367e51b4744d4c7eafb275d4782fb666)
  #15  pc 0x000000000009cc7c  /data/app/~~wsz9HcFa10baIQ8osrAiyA==/com.aimo.vika-Z8MMV_tqyIbL2M4UK-1bSw==/split_config.arm64_v8a.apk!libpag.so (BuildId: c43d6dd0367e51b4744d4c7eafb275d4782fb666)
  #16  pc 0x000000000009d7ac  /data/app/~~wsz9HcFa10baIQ8osrAiyA==/com.aimo.vika-Z8MMV_tqyIbL2M4UK-1bSw==/split_config.arm64_v8a.apk!libpag.so (BuildId: c43d6dd0367e51b4744d4c7eafb275d4782fb666)
  #17  pc 0x0000000000147940  /data/app/~~wsz9HcFa10baIQ8osrAiyA==/com.aimo.vika-Z8MMV_tqyIbL2M4UK-1bSw==/split_config.arm64_v8a.apk!libpag.so (BuildId: c43d6dd0367e51b4744d4c7eafb275d4782fb666)
  #18  pc 0x0000000000147c58  /data/app/~~wsz9HcFa10baIQ8osrAiyA==/com.aimo.vika-Z8MMV_tqyIbL2M4UK-1bSw==/split_config.arm64_v8a.apk!libpag.so (BuildId: c43d6dd0367e51b4744d4c7eafb275d4782fb666)
  #19  pc 0x00000000001483b8  /data/app/~~wsz9HcFa10baIQ8osrAiyA==/com.aimo.vika-Z8MMV_tqyIbL2M4UK-1bSw==/split_config.arm64_v8a.apk!libpag.so (BuildId: c43d6dd0367e51b4744d4c7eafb275d4782fb666)
  #20  pc 0x00000000000b8b98  /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+208)
  #21  pc 0x0000000000055794  /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+64)

有不少机器会有这个问题，这个是google后台反馈出来的，自己复现的话难以复现，
目前上报的android版本 android 12 和 android 13
手机的机型为 HONOR HNALI-Q  和 samsung a12s

**Comment by kevingpqi123** [comment, marked as answer]
1、首先要明确一个问题，不是所有的 pag 文件的性能都是一样的，如果遇到性能的问题，首先是排查素材，如果素材很复杂，首先去优化素材而不是其它的，如果素材很简单出现了卡顿，那才需要去排查是否是使用问题，一般而言，纯矢量的素材性能相对较好
https://pag.art/docs/optimization.html

业界没有可以兼容所有复杂素材的动效渲染方案，如果需要兼容效果，根本的解决办法就是分机型下发素材，不要期望复杂素材（含有 BMP 预合成的素材）可以在所有机型上没有卡顿问题

2、如果不涉及全屏播放、文本编辑、占位图替换和实色图层修改颜色，PAGView 可以切换至 PAGImageView，PAGImageView 支持缓存渲染结果，可以大幅降低概率（但不能从根本解决，首次渲染的时候仍有可能出现）
